### PR TITLE
CDAP-5191 FTP Batch Source implementation

### PR DESCRIPTION
**Pull request by gokulavasan**
FTP Batch Source and Unit test

JIRA : https://issues.cask.co/browse/CDAP-5191

### CODE REVIEW

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+  // Limit the number of splits to 1 since FTPInputStream does not support seek;
+  private static String limitSplits(@Nullable String fsProperties) {
+    Map<String, String> providedProperties;
+    if (fsProperties == null) {
+      providedProperties = new HashMap<>();
+    } else {
+      providedProperties = GSON.fromJson(fsProperties, MAP_STRING_STRING_TYPE);
+    }
+    providedProperties.put(FileInputFormat.SPLIT_MINSIZE, Long.toString(Long.MAX_VALUE));
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java]: in the comment you say limiting the splits to 1, but the value is Long.MAX_VALUE, shouldn't it be 1 ?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    public String fileRegex;
+
+    @Description(FileBatchSource.INPUT_FORMAT_CLASS_DESCRIPTION)
+    @Nullable
+    public String inputFormatClassName;
+
+    public FTPBatchSourceConfig() {
+      this.fileSystemProperties = GSON.toJson(ImmutableMap.<String, String>of());
+      this.fileRegex = ".*";
+    }
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java:88]: can you extract these properties out of FileBatchConfig, so you don't have to pass null for those two params as well in constructor.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java [hunk 3]
```diff
@@ -88,7 +88,7 @@
     "2015-06-16-15 (June 16th 2015, 3pm), it will read in files that contain '2015-06-16-14' in the filename. " +
     "If the field 'timeTable' is present, then it will read in files that have not yet been read. Defaults to '.*', " +
     "which indicates that no files will be filtered.";
-  private static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
+  protected static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
```
Comment (gsps1) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java:91]: why this change?

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+ * FTP Source Test.
+ */
+public class ETLFTPTestRun extends ETLBatchTestBase {
+
+  private static final String USER = "ftp";
+  private static final String PWD = "abcd";
+  private static final String TEST_STRING = "Hello World";
+
+  @ClassRule
+  public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();
```
Comment (gsps1) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java]: can you reuse the Temp folder in test base ?

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    ftpServer.setServerControlPort(0);
+
+    FileSystem fileSystem = new UnixFakeFileSystem();
+    fileSystem.add(new FileEntry(file.getAbsolutePath(), TEST_STRING));
+    ftpServer.setFileSystem(fileSystem);
+
+    ftpServer.addUserAccount(new UserAccount(USER, PWD, folder.getAbsolutePath()));
+    ftpServer.start();
+
+    Assert.assertTrue(ftpServer.isStarted());
```
Comment (gsps1) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java]: can this be false? if so, should we retry this check for it to be started

**Comment by gsps1** [comment]
what would happen in failure scenario's ?  path is incorrect, access errors, etc? would we return appropriate error messages?

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java (hunk 4 above)
Comment (gokulavasan) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java]: Yes I can. Will remove.

File: core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java (hunk 5 above)
Comment (gokulavasan) [inline core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java]: Yes, I will add a try mechanism to check

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java (hunk 3 above)
Comment (gokulavasan) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java:91]: So that we can access the property description in FTPBatchSourceConfig.

**Comment by gokulavasan** [comment]
@gsps1 Regarding the question about failure scenarios, we delegate to FileBatchSource. We don't perform any additional checks.

**Comment by gokulavasan** [comment]
Reopened here https://github.com/caskdata/hydrator-plugins/pull/182
